File: vgihan/baekjoon/나무자르기.js
const readline = require("readline");
const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

let n, m;

rl.on("line", function (line) {
  if (!n) {
    [n, m] = line.split(" ").map((v) => parseInt(v));
    return;
  }
  console.log(
    solution(
      line.split(" ").map((v) => parseInt(v)),
      m
    )
  );
  rl.close();
}).on("close", function () {
  process.exit();
});

const MAX_LENGTH = 2_000_000_000;

function solution(trees, m) {
  return search(0, MAX_LENGTH, trees, m);
}

function cutAllTrees(trees, height) {
  return trees.reduce((result, tree) => {
    return result + (tree > height ? tree - height : 0);
  }, 0);
}

function search(start, end, trees, m) {
  const mid = Math.ceil((start + end) / 2);
  const curGetTree = cutAllTrees(trees, mid);

  if (start >= end) return mid;
  if (curGetTree < m) return search(start, mid - 1, trees, m);

  return search(mid, end, trees, m);
}

File: thingseong/나무 자르기.py
import sys

N, M = map(int, sys.stdin.readline().split())

tree = list(map(int, sys.stdin.readline().split()))


s = 1
e = max(tree)




while s <= e:
    m = (s + e) // 2
    length = 0

    for t in tree:
        length += t - m if t-m > 0 else 0


    if length >= M:
        s = m + 1
    else:
        e = m - 1
print(e)
        
